tests/test_builders/test_build_linkcheck.py

```diff
@@ -99,9 +120,16 @@ def connection_count(self):
         return len(self.connections)
 
 
+@contextmanager
+def serve_sources(handler, rewrite_file: Path | None = None, tls_enabled: bool = False):
+    server_context = https_server if tls_enabled else http_server
+    with server_context(handler) as port, rewrite_internal_hyperlinks(rewrite_file, port):
```

Comment:
Perhaps instead of this somewhat-complicated-seeming additional contextmanager, it might be possible to rewrite these links using `linkcheck` builder event-handling.

The code used to handle rewriting of `github` links using that kind of approach: https://github.com/sphinx-doc/sphinx/blob/7894b5a2698c6abf4d285da78409785c40eab7b2/sphinx/builders/linkcheck.py#L684

I'll experiment with this to try to determine whether it's feasible, and if it is, whether it makes the test code more readable or not.

Reply:
It is possible, and I think it makes the resulting code cleaner.  #12144 would help too.